### ISSUE TITLE
tar: buildout-cache/eggs/setuptools-0.6c11-py2.7.egg: Cannot open: File exists

### ISSUE THREAD

**Issue by djowett** [collective/buildout.plonetest#18]
I got this issue on plone.app.multilingual & though I've got past it there, wonder if it might affect other packages using buildout-cache.cfg from here.

The error on Travis is:

    Installing download.
    Downloading https://launchpad.net/plone/4.3/4.3.2/+download/Plone-4.3.2-UnifiedInstaller.tgz
    download: Extracting package to /home/travis/build/plone/plone.app.multilingual/parts/download
    Installing install.
    tar: buildout-cache/eggs/setuptools-0.6c11-py2.7.egg: Cannot open: File exists
    tar: Exiting with failure status due to previous errors
    While:
      Installing install.

full travis log is [here](https://travis-ci.org/plone/plone.app.multilingual/jobs/92060436)

basically it seems python bootstrap was putting setuptools etc in buildout-cache/eggs/ and then tar xv doesn't handle overwriting files (at least on Ubuntu 12.04), so I [rewrote the install part](https://github.com/plone/plone.app.multilingual/blob/31aa9cc865ffe43292281a550233035b99aeb12d/test-plone-4.1.x.cfg#L10).  Does this seem useful here?

**Comment by mauritsvanrees** [comment]
The problem is that the setuptools egg is first a directory and the tarball has it as a file. I don't have this locally, but this can probably differ for various reasons.  When extracting a tarball, an empty directory will be replaced, but not a filled directory:

```
maurits@MauritsBook:test $ touch foo
maurits@MauritsBook:test $ tar czf test.tgz foo
maurits@MauritsBook:test $ tar xzf test.tgz
maurits@MauritsBook:test $ rm foo && mkdir foo
maurits@MauritsBook:test $ tar xzf test.tgz
maurits@MauritsBook:test $ ls -lR foo
-rw-r--r--  1 maurits  staff  0 20 Nov 21:18 foo
maurits@MauritsBook:test $ rm foo && mkdir foo && touch foo/bar
maurits@MauritsBook:test $ tar xzf test.tgz
tar: foo: Cannot open: File exists
tar: Exiting with failure status due to previous errors
```

There are a few tar options we can try, see http://www.gnu.org/software/tar/manual/html_node/Dealing-with-Old-Files.html

```
maurits@MauritsBook:test $ tar xzf test.tgz --overwrite
tar: foo: Cannot open: Is a directory
tar: Exiting with failure status due to previous errors
maurits@MauritsBook:test $ tar xzf test.tgz --overwrite-dir
tar: foo: Cannot open: File exists
tar: Exiting with failure status due to previous errors
maurits@MauritsBook:test $ tar xzf test.tgz --skip-old-files
maurits@MauritsBook:test $ ls -lR foo
total 0
-rw-r--r--  1 maurits  staff  0 20 Nov 21:19 bar
maurits@MauritsBook:test $ tar xzf test.tgz --unlink-first
tar: foo: Cannot unlink: Directory not empty
tar: Exiting with failure status due to previous errors
maurits@MauritsBook:test $ tar xzf test.tgz --recursive-unlink
maurits@MauritsBook:test $ ls -lR foo
-rw-r--r--  1 maurits  staff  0 20 Nov 21:18 foo
```

So:

- `--skip-old-files` skips existing directories/files/symlinks without complaining: they are not overridden.

-  `-recursive-unlink` removes whatever of the same name is there (filled directory, file, symlink) and replaces it with the file from the archive.

`--skip-old-files` may be the best one here.

**Comment by djowett** [comment]
Ah,but beware... Travis is (mostly) using ubuntu 12.04, and tar on that doesn't have all the options, including iirc "--skip-old-files"

**Comment by mauritsvanrees** [comment]
Indeed.  I already fixed it in d891ae3a1aa86572397466326789d73e3873ac97, using your strategy.
